Improve status update error handling in UserAccount controller

alexeykazakov opened the pull request:
Followup on https://github.com/codeready-toolchain/member-operator/pull/20/files#r297629374

- Use logger with values to log messages/errors
- Set status to "provisioning" before creating a new user or identity (not after)
- No error ignoring

TODO:
- [x] Tests. I wonder if we can use some fake `reconcile.Request` to test the entire `Reconcile()` method...

nurali-techie commented on pkg/controller/useraccount/useraccount_controller.go:
```diff
@@ -96,92 +97,90 @@ func (r *ReconcileUserAccount) Reconcile(request reconcile.Request) (reconcile.R
 	var user *userv1.User
-	if user, created, err = r.ensureUser(userAcc); err != nil {
-		_ = r.updateStatus(userAcc, toolchainv1alpha1.StatusProvisioning, err.Error())
+	if user, err = r.ensureUser(reqLogger, userAcc); err != nil {
```
When I start operator locally and then try to create user_acocunt .. I am getting below error .. I guess this is due to race_condition that we fix before with introducing `created` flag.
```
{"level":"info","ts":1561608964.0394926,"logger":"controller_useraccount","msg":"Reconciling UserAccount","Request.Namespace":"toolchain-member-operator","Request.Name":"johnsmith"}
{"level":"info","ts":1561608964.0395193,"logger":"controller_useraccount","msg":"creating a new user","Request.Namespace":"toolchain-member-operator","Request.Name":"johnsmith","name":"johnsmith"}
{"level":"info","ts":1561608964.050983,"logger":"controller_useraccount","msg":"user created successfully","Request.Namespace":"toolchain-member-operator","Request.Name":"johnsmith","name":"johnsmith"}
{"level":"info","ts":1561608964.0510187,"logger":"controller_useraccount","msg":"creating a new identity","Request.Namespace":"toolchain-member-operator","Request.Name":"johnsmith","name":"rhd:1a03ecac-7c0b-44fc-b66d-12dd7fb21c40"}
{"level":"info","ts":1561608964.0628774,"logger":"controller_useraccount","msg":"identity created successfully","Request.Namespace":"toolchain-member-operator","Request.Name":"johnsmith","name":"rhd:1a03ecac-7c0b-44fc-b66d-12dd7fb21c40"}
{"level":"info","ts":1561608964.077371,"logger":"controller_useraccount","msg":"user-identity mapping created successfully","Request.Namespace":"toolchain-member-operator","Request.Name":"johnsmith","name":"rhd:1a03ecac-7c0b-44fc-b66d-12dd7fb21c40"}
{"level":"info","ts":1561608964.1616054,"logger":"controller_useraccount","msg":"Reconciling UserAccount","Request.Namespace":"","Request.Name":"johnsmith"}
{"level":"info","ts":1561608964.1617534,"logger":"controller_useraccount","msg":"creating a new user","Request.Namespace":"","Request.Name":"johnsmith","name":"johnsmith"}
{"level":"error","ts":1561608964.2006922,"logger":"kubebuilder.controller","msg":"Reconciler error","controller":"useraccount-controller","request":"/johnsmith","error":"failed to create user '[johnsmith]': users.user.openshift.io \"johnsmith\" already exists","errorVerbose":"users.user.openshift.io \"johnsmith\" already exists\nfailed to create user '[johnsmith]'\ngithub.com/codeready-toolchain/member-operator/pkg/controller/useraccount.(*ReconcileUserAccount).wrapErrorWithStatusUpdate\n\t/home/nvirani/go/src/github.com/codeready-toolchain/member-operator/pkg/controller/useraccount/useraccount_controller.go:203\ngithub.com/codeready-toolchain/member-operator/pkg/controller/useraccount.(*ReconcileUserAccount).ensureUser\n\t/home/nvirani/go/src/github.com/codeready-toolchain/member-operator/pkg/controller/useraccount/useraccount_controller.go:131\ngithub.com/codeready-toolchain/member-operator/pkg/controller/useraccount.(*ReconcileUserAccount).Reconcile\n\t/home/nvirani/go/src/github.com/codeready-toolchain/member-operator/pkg/controller/useraccount/useraccount_controller.go:101\ngithub.com/codeready-toolchain/member-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/home/nvirani/go/src/github.com/codeready-toolchain/member-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:215\ngithub.com/codeready-toolchain/member-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1\n\t/home/nvirani/go/src/github.com/codeready-toolchain/member-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158\ngithub.com/codeready-toolchain/member-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/home/nvirani/go/src/github.com/codeready-toolchain/member-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133\ngithub.com/codeready-toolchain/member-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/home/nvirani/go/src/github.com/codeready-toolchain/member-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134\ngithub.com/codeready-toolchain/member-operator/vendor/k8s.io/apimachinery/pkg/util/wait.Until\n\t/home/nvirani/go/src/github.com/codeready-toolchain/member-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1337","stacktrace":"github.com/codeready-toolchain/member-operator/vendor/github.com/go-logr/zapr.(*zapLogger).Error\n\t/home/nvirani/go/src/github.com/codeready-toolchain/member-operator/vendor/github.com/go-logr/zapr/zapr.go:128\ngithub.com/codeready-toolchain/member-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/home/nvirani/go/src/github.com/codeready-toolchain/member-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:217\ngithub.com/codeready-toolchain/member-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1\n\t/home/nvirani/go/src/github.com/codeready-toolchain/member-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158\ngithub.com/codeready-toolchain/member-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/home/nvirani/go/src/github.com/codeready-toolchain/member-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133\ngithub.com/codeready-toolchain/member-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/home/nvirani/go/src/github.com/codeready-toolchain/member-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134\ngithub.com/codeready-toolchain/member-operator/vendor/k8s.io/apimachinery/pkg/util/wait.Until\n\t/home/nvirani/go/src/github.com/codeready-toolchain/member-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88"}
{"level":"info","ts":1561608965.2020411,"logger":"controller_useraccount","msg":"Reconciling UserAccount","Request.Namespace":"","Request.Name":"johnsmith"}
{"level":"info","ts":1561608965.202231,"logger":"controller_useraccount","msg":"user already exists","Request.Namespace":"","Request.Name":"johnsmith","name":"johnsmith"}
{"level":"info","ts":1561608965.2023094,"logger":"controller_useraccount","msg":"identity already exists","Request.Namespace":"","Request.Name":"johnsmith","name":"rhd:1a03ecac-7c0b-44fc-b66d-12dd7fb21c40"}
{"level":"info","ts":1561608965.2174907,"logger":"controller_useraccount","msg":"user-identity mapping already exists","Request.Namespace":"","Request.Name":"johnsmith","name":"rhd:1a03ecac-7c0b-44fc-b66d-12dd7fb21c40"}
```

alexeykazakov replied:
Yes. This is because of the removing the flag. I caught that bug too during writing tests. Will fix it shortly.

alexeykazakov replied:
Fixed + added tests.

alexeykazakov commented:
I added tests. Ready for review.

nurali-techie commented on pkg/controller/useraccount/useraccount_controller.go:
```diff
@@ -98,75 +99,74 @@ func (r *ReconcileUserAccount) Reconcile(request reconcile.Request) (reconcile.R
+				err = r.client.Create(context.TODO(), identity)
+			}
+			if err != nil {
+				return nil, false, r.wrapErrorWithStatusUpdate(logger, userAcc, err, "failed to create identity '%s'", name)
```
This doesn't look obvious error_handling.  In case `SetControllerReference` returns error.  First we are testing `err == nil` and then we are testing `err != nil`.  It's testing err values two times instead of one.
So overall, code looks fine but execution suffers here.

alexeykazakov replied:
Yes, it’s because err can become not nil after the first check. The code is a bit ambiguous I agree. Alternatively we could check two errors independently which will make the code a bit more verbose but could make it easier to read.

nurali-techie replied:
This is less go_ideomatic way. go says first check error and then proceed.  I really need to execute the code in my mind to understand it.  With that input, I will leave it up to you :+1:

alexeykazakov replied:
Addressed in https://github.com/codeready-toolchain/member-operator/pull/25/commits/81f9d57c0d73f11d0430e1ae57b28891038d1a96

nurali-techie commented on pkg/controller/useraccount/useraccount_controller.go:
```diff
@@ -98,75 +99,74 @@ func (r *ReconcileUserAccount) Reconcile(request reconcile.Request) (reconcile.R
-			identity = newIdentity(userAcc)
-			if err := controllerutil.SetControllerReference(userAcc, identity, r.scheme); err != nil {
+			logger.Info("creating a new identity", "name", name)
+			if err := r.updateStatus(userAcc, toolchainv1alpha1.StatusProvisioning, err.Error()); err != nil {
```
For happy scenario (first time reconcile for new user_account), this `updateStatus()` will set status=provisioning and error="user not found".  Assuming rest of the thing went well for create user and we returned.  Now, the status.error is still the same which is wrong.  We need to reset error to blank.  Below is the log where I am printing status on start of each reconcile().  You can see in second reconcile the status="provisioning" and error="user not found" but first reconcile run successfully with creating user.

```
{"level":"info","ts":1561619157.6916687,"logger":"controller_useraccount","msg":"Reconciling UserAccount","Request.Namespace":"toolchain-member-operator","Request.Name":"johnsmith"}
VN: u.s= { }
{"level":"info","ts":1561619157.6917093,"logger":"controller_useraccount","msg":"creating a new user","Request.Namespace":"toolchain-member-operator","Request.Name":"johnsmith","name":"johnsmith"}
{"level":"info","ts":1561619157.7047453,"logger":"controller_useraccount","msg":"user created successfully","Request.Namespace":"toolchain-member-operator","Request.Name":"johnsmith","name":"johnsmith"}
{"level":"info","ts":1561619157.8219233,"logger":"controller_useraccount","msg":"Reconciling UserAccount","Request.Namespace":"","Request.Name":"johnsmith"}
VN: u.s= {provisioning User.user.openshift.io "johnsmith" not found}
{"level":"info","ts":1561619157.8219843,"logger":"controller_useraccount","msg":"user already exists","Request.Namespace":"","Request.Name":"johnsmith","name":"johnsmith"}
{"level":"info","ts":1561619157.8220155,"logger":"controller_useraccount","msg":"creating a new identity","Request.Namespace":"","Request.Name":"johnsmith","name":"rhd:1a03ecac-7c0b-44fc-b66d-12dd7fb21c40"}
{"level":"info","ts":1561619157.8332338,"logger":"controller_useraccount","msg":"identity created successfully","Request.Namespace":"","Request.Name":"johnsmith","name":"rhd:1a03ecac-7c0b-44fc-b66d-12dd7fb21c40"}
{"level":"info","ts":1561619158.0225096,"logger":"controller_useraccount","msg":"Reconciling UserAccount","Request.Namespace":"","Request.Name":"johnsmith"}
```

alexeykazakov replied:
Good catch. I will address that and will add a test for that.

alexeykazakov replied:
Addressed in https://github.com/codeready-toolchain/member-operator/pull/25/commits/81f9d57c0d73f11d0430e1ae57b28891038d1a96

MatousJobanek commented on pkg/controller/useraccount/useraccount_controller.go:
```diff
@@ -98,75 +99,74 @@ func (r *ReconcileUserAccount) Reconcile(request reconcile.Request) (reconcile.R
-			user = newUser(userAcc)
-			if err := controllerutil.SetControllerReference(userAcc, user, r.scheme); err != nil {
+			logger.Info("creating a new user", "name", userAcc.Name)
+			if err := r.updateStatus(userAcc, toolchainv1alpha1.StatusProvisioning, err.Error()); err != nil {
```
do we really need to set the error here? I think that it is obvious that the `User` does/did not exist when the status is `provisioning`. I would use `error` field only when something fails.

MatousJobanek replied:
ah, I just figured out that the Nurali's [comment](https://github.com/codeready-toolchain/member-operator/pull/25/files#r298037991) is about the same :-)

alexeykazakov replied:
Addressed in https://github.com/codeready-toolchain/member-operator/pull/25/commits/81f9d57c0d73f11d0430e1ae57b28891038d1a96

MatousJobanek commented on pkg/controller/useraccount/useraccount_controller.go:
```diff
@@ -98,75 +99,74 @@ func (r *ReconcileUserAccount) Reconcile(request reconcile.Request) (reconcile.R
 			return user, true, nil
 		}
-		return nil, false, errs.Wrapf(err, "failed to get user '%s'", userAcc.Name)
+		return nil, false, r.wrapErrorWithStatusUpdate(logger, userAcc, err, "failed to get user '%s'", userAcc.Name)
```
In this case, the `wrapErrorWithStatusUpdate` sets the status = `provisioning` and the error to the error msg, but the reconcile doesn't have to be for `POST` request, but for `PATCH`. This means that the status should be rather `updating` instead of `provisioning`. That's why I don't like this automatic setting of the status to `provisioning` when something fails - I would rather use either `failed`,`failing` or `unknown`

alexeykazakov replied:
This is something to discuss...
Let try to look at other examples.

MatousJobanek replied:
according to our discussion either here or in slack channel, it's a bit more complex task and should be handled in separated PR - could you please create a Jira for that?

alexeykazakov replied:
https://jira.coreos.com/browse/CRT-166

MatousJobanek commented on pkg/controller/useraccount/useraccount_controller_test.go:
```diff
@@ -1,134 +1,163 @@
-			scheme: s,
-		}
+	// Register operator types with the runtime scheme.
+	s.AddKnownTypes(toolchainv1alpha1.SchemeGroupVersion, userAcc)
```
we should either use this way of adding types to the scheme or using the `apis` package https://github.com/codeready-toolchain/member-operator/pull/25/files#diff-fc74d9d211758dbe5bd817320a8fe2a1R29 
using both of them is not really needed I guess.

alexeykazakov replied:
Addressed in https://github.com/codeready-toolchain/member-operator/pull/25/commits/81f9d57c0d73f11d0430e1ae57b28891038d1a96

MatousJobanek commented on pkg/controller/useraccount/useraccount_controller_test.go:
```diff
@@ -1,134 +1,163 @@
-		assert.Equal(t, userAcc.Name, newUser.Name)
-	})
-}
+	client := fake.NewFakeClient(objs...)
```
I think that you don't need the `objs` array - you can pass the object directly to the function:
```go
client := fake.NewFakeClient(userAcc)
```

alexeykazakov replied:
Addressed in https://github.com/codeready-toolchain/member-operator/pull/25/commits/81f9d57c0d73f11d0430e1ae57b28891038d1a96

MatousJobanek commented on pkg/controller/useraccount/useraccount_controller_test.go:
```diff
@@ -1,134 +1,163 @@
+		// given
+		// No user account exists
+		objs := []runtime.Object{}
+		client := fake.NewFakeClient(objs...)
```
In this case you don't need to pass to the function anything as the array doesn't contain any object anyway:
```go
client := fake.NewFakeClient()
```

alexeykazakov replied:
Addressed in https://github.com/codeready-toolchain/member-operator/pull/25/commits/81f9d57c0d73f11d0430e1ae57b28891038d1a96

MatousJobanek commented on pkg/controller/useraccount/useraccount_controller_test.go:
```diff
@@ -1,134 +1,163 @@
-		assert.NoError(t, err)
-		assert.False(t, created)
-		assert.NotNil(t, newIdentity)
+	t.Run("create_identity", func(t *testing.T) {
```
this subtest setup (and also the following ones) is a bit problematic because it relies on the previous subtest - it expects that the User already exists, so it's no possible to run it separately.

alexeykazakov replied:
That's correct. These subtests are different steps of the same test rather than independent tests. I still prefer to keep them as is. Otherwise the test will have unnecessarily duplicated code. I can add comments though...

alexeykazakov replied:
Added comments in https://github.com/codeready-toolchain/member-operator/pull/25/commits/81f9d57c0d73f11d0430e1ae57b28891038d1a96

MatousJobanek replied:
But then we are losing the big benefit of subtests - to be able to run them separately and when one subtest fails, then the other dependent ones should not be affected.
I'm not a fan of the (sub)tests that depend on previous ones (if not really needed as can be for example in UI tests, but this is not the case).  
As for your concern about duplicated code - I don't think that there would be too much of duplication. I tried to quickly rewrite the code - see here https://gist.github.com/MatousJobanek/4816087ca5f69df0434290086f022d14 could be still improved

alexeykazakov replied:
IMO it's really not a big deal in this case. But anyway.. here we go - https://github.com/codeready-toolchain/member-operator/pull/25/commits/358de797d11bab64e776286f73433ceba39ff619

MatousJobanek commented on pkg/controller/useraccount/useraccount_controller_test.go:
```diff
@@ -1,134 +1,163 @@
+		updatedAcc := &toolchainv1alpha1.UserAccount{}
+		err = r.client.Get(context.TODO(), types.NamespacedName{Namespace: req.Namespace, Name: userAcc.Name}, updatedAcc)
+		require.NoError(t, err)
+		assert.Equal(t, toolchainv1alpha1.StatusProvisioning, updatedAcc.Status.Status)
```
I think that the `Status.Error` field should be verified as well

alexeykazakov replied:
Addressed in https://github.com/codeready-toolchain/member-operator/pull/25/commits/81f9d57c0d73f11d0430e1ae57b28891038d1a96

alexeykazakov commented on pkg/controller/useraccount/useraccount_controller.go:
```diff
@@ -98,75 +99,74 @@ func (r *ReconcileUserAccount) Reconcile(request reconcile.Request) (reconcile.R
-		return reconcile.Result{}, r.updateStatus(userAcc, toolchainv1alpha1.StatusProvisioning, "")
 	}
 
 	return reconcile.Result{}, r.updateStatus(userAcc, toolchainv1alpha1.StatusProvisioned, "")
```
I actually added such a check originally but then removed it. I was thinking that maybe the client already takes care of it and do not set patch requests if there is nothing to change. Will need to do additional research. So, decided not to change it until we know for sure.

alexeykazakov replied:
I did some research. The client doesn't seem to check the actual changes and sends `PUT` request. So, I'm adding the check - https://github.com/codeready-toolchain/member-operator/pull/25/commits/1bcaee4825a7e6b068e8bc3137822ef9618a2641

alexeykazakov commented:
@nurali-techie @MatousJobanek thanks for the review. I addressed most of the comments. Only two comments from Matous are left untouched since they are not introduced by this PR and I would like to do additional research before changing anything. Let's address them in separate PRs if needed.

alexeykazakov commented:
/lgtm

xcoulon commented:
/lgtm

alexeykazakov commented:
/test images

alexeykazakov commented:
/refresh
